### ISSUE TITLE
`as` removes bases styles for functional/class components

### ISSUE THREAD

**Issue by trevorwright** [styled-components/styled-components#2375]
## System:
 - OS: macOS 10.14.2
 - Shell: 5.3 - /bin/zsh
## Binaries:
 - Node: 8.9.0
 - Yarn: 1.12.3
## npmPackages:
 - babel-plugin-styled-components: 1.5.1 => 1.5.1
 - styled-components: 4.1.2 => 4.1.2

## Reproduction
https://codesandbox.io/s/30k0n99jv5

## Steps to reproduce
1) Create Styled-Component
2) Use the Styled-Component in a functional component (be sure to pass down your props to the styled-component)
3) Extend functional component with some additional styles
4) Use your extended version of the component with the `as` as prop

## Expected Behavior
* Base styles should still be applied
* Extended styles should be applied overtop of base styles
* New tag is used

## Actual Behavior
* Base styles are dropped, and only the extended styles are applied
* New tag is still properly applied.

**Comment by AustP** [comment]
I'm seeing this on 4.1.3 as well.

Headline.js:

    import styled from 'styled-components';

    const Headline = styled.h1`
      color: orange;
      display: block;
      font-weight: 300;
    `;

    export default Headline;

StyledHeadline.js:

    import { Headline } from 'components';
    import styled from 'styled-components';

    const StyledHeadline = styled(Headline)`
      color: pink;
    `;

When I use `StyledHeadline` with the `as` prop, the only styles applied are `color: pink;`.

**Comment by jole78** [comment]
I'm seeing the same thing when a base component has some default props...those aren't taken into consideration either when using the as prop on a child component.

**Comment by chasemccoy** [comment]
I too am seeing this issue on 4.1.3

**Comment by probablyup** [comment]
The reason this happens is because `as` switches out the underlying component being rendered. If the target wrapped in `styled` is not another styled component, passing `as` will retain deeper styles since there is a custom component layer between, breaking the ability to "fold" the components.

**Comment by probablyup** [comment]
If you do want to retain the ability to fold the styles as deeply as possible, I'd suggest wrapping your custom components in `styled()` and exporting that.

```jsx
function FooBase(props) {
  return <div {...props} />;
}

export const Foo = styled(FooBase)`
  color: red;
`;
```

vs

```jsx
export function Foo(props) {
  return <Wrapper {...props} />;
}

const Wrapper = styled.div`
  color: red;
`;
```

**Comment by AustP** [comment]
@probablyup In my above Headline example, the target wrapped in `styled` *is* a styled component. I am passing the return value from `styled.h1` into the next `styled` call with no changes and still the base styles are mssing. Is there something else going on?

**Comment by probablyup** [comment]
@AustP Yours looks like an actual bug. If you can reproduce in a sandbox, can take a look

**Comment by thibautRe** [comment]
Hello,

I did a little bit of digging. A possible reason why this was reported as a bug (the original issue, not @AustP's variant) is because _it worked_ in the `4.0.0-beta` versions of SC, and stopped working at some point.

Here's a reproduction with [`sc@v4.0.0-beta.9`](https://codesandbox.io/s/pwmllzl3qj) where the bug doesn't happen - styled props of the inner component are kept and the outer component is a button.
With [`sc@v4.0.0-beta.10`](https://codesandbox.io/s/7j37zjmmv6) though, the styles of the inner component are lost.

So it happened [somewhere in that diff](https://github.com/styled-components/styled-components/compare/v4.0.0-beta.9..v4.0.0-beta.10)

I will try to keep doing some digging - from my side, it prevents me from moving away from the beta version of SC because there is no clean way around.

**Comment by leerob** [comment]
- Another example: https://codesandbox.io/s/r7kjpyjwq
- Related issue: https://github.com/styled-components/styled-components/issues/1981

What is the best workaround in the meantime?

**Comment by usagi-f** [comment]
I encountered this issue too.

In case of use styled-component for the UI-library that has internally gives className, and using `as` option will behave exactly as the pasted by guys the examples.
This could not keep the appearance as enough UI component.

I read styled-component's source code a little.
For now, I just found behavior is below.

```js
const elementToBeCreated = this.props.as || this.attrs.as || target;
```
https://github.com/styled-components/styled-components/blob/master/packages/styled-components/src/models/StyledComponent.js#L135

```js
return createElement(elementToBeCreated, propsForElement);
```
https://github.com/styled-components/styled-components/blob/master/packages/styled-components/src/models/StyledComponent.js#L172

Is it generated another element when use `as`?
Which means, Even if can be passed enough props object by `propsForElement`, the internal logic by the component itself that is originally passed as a `target` is not executed.

If the classNames within a component are internally generated by different architectures such as `classnames`, the component will be destroyed when use with `as`.

Is `as` use case only use to the primitive element?

This is might become a stall point when use styled-component with some UI library.
Also, this may be breaking if the UI library has props of the same name (`as`).

If my understanding is not wrong, the best workaround is ... "Take care when use `as` option" 🤕
Can I know the contributor's opinion about this?

cc: @probablyup 
@leerob thanks for your example, It is easy to understand.
